cli/cmd/create.go

```diff
@@ -48,259 +33,69 @@ var createCmd = &cobra.Command{
 	Short:        "Create new stack",
 	Long:         "Create a new stack with a given tag.",
 	SilenceUsage: true,
-	PreRunE:      happyCmd.Validate(checkCreateFlags, cobra.ExactArgs(1), happyCmd.CheckStackName),
-	RunE:         runCreate,
-}
-
-func checkCreateFlags(cmd *cobra.Command, args []string) error {
-	if cmd.Flags().Changed("skip-check-tag") && !cmd.Flags().Changed("tag") {
-		return errors.New("--skip-check-tag can only be used when --tag is specified")
-	}
-
-	if !createTag && !cmd.Flags().Changed("tag") {
-		return errors.New("Must specify a tag when create-tag=false")
-	}
-
-	return nil
+	PreRunE: happyCmd.Validate(
+		happyCmd.IsTagUsedWithSkipTag,
+		cobra.ExactArgs(1),
+		happyCmd.IsStackNameDNSCharset,
+		happyCmd.IsStackNameAlphaNumeric),
+	RunE: runCreate,
 }
 
-func runCreate(cmd *cobra.Command, args []string) error {
+func runCreate(
+	cmd *cobra.Command,
+	args []string,
+) error {
 	ctx := cmd.Context()
-	stackName := args[0]
-
-	isDryRun := util.DryRunType(dryRun)
-
-	if !regexp.MustCompile(`^[a-z0-9\-]*$`).MatchString(stackName) {
-		return errors.New("Stack name must contain only lowercase letters, numbers, and hyphens/dashes")
-	}
-
-	bootstrapConfig, err := config.NewBootstrapConfig(cmd)
-	if err != nil {
-		return err
-	}
-	happyConfig, err := config.NewHappyConfig(bootstrapConfig)
-	if err != nil {
-		return err
-	}
-
-	backend, err := backend.NewAWSBackend(ctx, happyConfig)
-	if err != nil {
-		return err
-	}
-
-	builderConfig := artifact_builder.NewBuilderConfig().WithBootstrap(bootstrapConfig).WithHappyConfig(happyConfig).WithDryRun(isDryRun)
-
-	// slice support parity with update command
-	buildOpts := []artifact_builder.ArtifactBuilderBuildOption{}
-	// FIXME: this is an error-prone interface
-	// if slice specified, use it
-	if sliceName != "" {
-		slice, err := happyConfig.GetSlice(sliceName)
-		if err != nil {
-			return err
-		}
-		buildOpts = append(buildOpts, artifact_builder.BuildSlice(slice))
-		builderConfig.WithProfile(slice.Profile)
-	}
-	ab := artifact_builder.NewArtifactBuilder(isDryRun).WithConfig(builderConfig).WithBackend(backend)
-
-	workspaceRepo := createWorkspaceRepo(isDryRun, backend)
-
-	err = verifyTFEBacklog(ctx, workspaceRepo)
+	happyConfig, stackService, artifactBuilder, tag, stackTags, awsBackend, err := initializeHappyClients(
+		cmd,
```

Comment:
I feel like the number of arguments this method returns is excessive, there's a lot of implied complexity in this approach, it's better to wrap happyConfig, stackService, artifactBuilder, tag, stackTags, awsBackend into a separate struct and return thatstruct, err.

Reply:
👍🏻 ya agreed